File: node/crypto/channel/x3dh.go
package channel

import (
	"crypto/sha512"

	"golang.org/x/crypto/hkdf"
	"source.quilibrium.com/quilibrium/monorepo/nekryptology/pkg/core/curves"
)

var domainSeparators = map[string][]byte{
	curves.ED448().Name: {
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
		0xFF,
	},
}

func SenderX3DH(
	sendingIdentityPrivateKey curves.Scalar,
	sendingEphemeralPrivateKey curves.Scalar,
	receivingIdentityKey curves.Point,
	receivingSignedPreKey curves.Point,
	sessionKeyLength uint8,
) []byte {
	xdh1 := receivingSignedPreKey.Mul(
		sendingIdentityPrivateKey,
	).ToAffineCompressed()
	xdh2 := receivingIdentityKey.Mul(
		sendingEphemeralPrivateKey,
	).ToAffineCompressed()
	xdh3 := receivingSignedPreKey.Mul(
		sendingEphemeralPrivateKey,
	).ToAffineCompressed()
	salt := make([]byte, sessionKeyLength)
	x3dh := hkdf.New(sha512.New, append(
		append(
			append(
				append([]byte{}, domainSeparators[receivingIdentityKey.CurveName()]...),
				xdh1[:]...),
			xdh2[:]...),
		xdh3[:]...), salt, []byte("quilibrium-x3dh"))
	sessionKey := make([]byte, sessionKeyLength)
	if _, err := x3dh.Read(sessionKey[:]); err != nil {
		return nil
	}

	return sessionKey
}

func ReceiverX3DH(
	sendingIdentityPrivateKey curves.Scalar,
	sendingSignedPrePrivateKey curves.Scalar,
	receivingIdentityKey curves.Point,
	receivingEphemeralKey curves.Point,
	sessionKeyLength uint8,
) []byte {
	xdh1 := receivingIdentityKey.Mul(
		sendingSignedPrePrivateKey,
	).ToAffineCompressed()
	xdh2 := receivingEphemeralKey.Mul(
		sendingIdentityPrivateKey,
	).ToAffineCompressed()
	xdh3 := receivingEphemeralKey.Mul(
		sendingSignedPrePrivateKey,
	).ToAffineCompressed()
	salt := make([]byte, sessionKeyLength)
	x3dh := hkdf.New(sha512.New, append(
		append(
			append(
				append([]byte{}, domainSeparators[receivingIdentityKey.CurveName()]...),
				xdh1[:]...),
			xdh2[:]...),
		xdh3[:]...), salt, []byte("quilibrium-x3dh"))
	sessionKey := make([]byte, sessionKeyLength)
	if _, err := x3dh.Read(sessionKey[:]); err != nil {
		return nil
	}

	return sessionKey
}

File: crates/channel/src/protocols/x3dh.rs
use std::collections::HashMap;
use sha2::Sha512;
use hkdf::Hkdf;
use ed448_goldilocks_plus::{subtle, CompressedEdwardsY, EdwardsPoint, Scalar};
use lazy_static::lazy_static;

lazy_static! {
    static ref DOMAIN_SEPARATORS: HashMap<&'static str, Vec<u8>> = {
        let mut m = HashMap::new();
        m.insert("ed448", vec![
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
            0xFF,
        ]);
        m
    };
}

pub fn sender_x3dh(
    sending_identity_private_key: &Scalar,
    sending_ephemeral_private_key: &Scalar,
    receiving_identity_key: &EdwardsPoint,
    receiving_signed_pre_key: &EdwardsPoint,
    session_key_length: usize,
) -> Option<Vec<u8>> {
    let xdh1 = (receiving_signed_pre_key * sending_identity_private_key).compress().to_bytes().to_vec();
    let xdh2 = (receiving_identity_key * sending_ephemeral_private_key).compress().to_bytes().to_vec();
    let xdh3 = (receiving_signed_pre_key * sending_ephemeral_private_key).compress().to_bytes().to_vec();

    let salt = vec![0u8; session_key_length];
    let info = b"quilibrium-x3dh";

    let domain_separator = DOMAIN_SEPARATORS.get("ed448")
        .expect("Unsupported curve");

    let mut ikm = Vec::<u8>::new();
    ikm.extend(domain_separator);
    ikm.extend(xdh1);
    ikm.extend(xdh2);
    ikm.extend(xdh3);

    let hk = Hkdf::<Sha512>::new(Some(&salt), &ikm);
    let mut session_key = vec![0u8; session_key_length];
    hk.expand(info, &mut session_key).ok()?;

    Some(session_key)
}

pub fn receiver_x3dh(
    sending_identity_private_key: &Scalar,
    sending_signed_pre_private_key: &Scalar,
    receiving_identity_key: &EdwardsPoint,
    receiving_ephemeral_key: &EdwardsPoint,
    session_key_length: usize,
) -> Option<Vec<u8>> {
    let xdh1 = (receiving_identity_key * sending_signed_pre_private_key).compress().to_bytes().to_vec();
    let xdh2 = (receiving_ephemeral_key * sending_identity_private_key).compress().to_bytes().to_vec();
    let xdh3 = (receiving_ephemeral_key * sending_signed_pre_private_key).compress().to_bytes().to_vec();

    let salt = vec![0u8; session_key_length];
    let info = b"quilibrium-x3dh";

    let domain_separator = DOMAIN_SEPARATORS.get("ed448")
        .expect("Unsupported curve");

    let mut ikm = Vec::<u8>::new();
    ikm.extend(domain_separator);
    ikm.extend(xdh1);
    ikm.extend(xdh2);
    ikm.extend(xdh3);

    let hk = Hkdf::<Sha512>::new(Some(&salt), &ikm);
    let mut session_key = vec![0u8; session_key_length];
    hk.expand(info, &mut session_key).ok()?;

    Some(session_key)
}